### DISCUSSION TITLE
Trigger events in testing scenarios

### DISCUSSION THREAD

**Discussion by abingham** [bigskysoftware/htmx#2921] [Q&A] [answered]
Use use httpx.Client in my tests to make sure my endpoints all work as expected. Naively, I had expected it to help test various HTMX features in the returned pages. For example, I use `hx-trigger="load"` and `hx-trigger="revealed"` in some places, and I thought these would be triggered. In hindsight, of course, this was a wrongheaded expectation, but it left me wondering what I should do to actually exercise these triggers.

Do I need to go with a full selenium-style test, i.e. with an actual browser, to interact with these triggers? Or is there some lighter-weight "small headless browser" I can use to fetch my pages and render them? Or perhaps some other kind of approach I just don't know about?

Any advice, suggestions, or warnings would be great!

**Comment by Telroshan** [comment, marked as answer]
>  Or is there some lighter-weight "small headless browser" I can use to fetch my pages and render them

That's what we do for htmx's own test suite, we use `mocha-chrome`.
See these [`hx-trigger="revealed"` tests](https://github.com/bigskysoftware/htmx/blob/master/test/attributes/hx-trigger.js#L864), or this [`hx-trigger="load"` one](https://github.com/bigskysoftware/htmx/blob/master/test/core/ajax.js#L255) for examples

Hope this helps!

**Comment by abingham** [comment]
Thanks, that does help a lot!
